Add a CanonicalOSSALifetime utility

atrick opened the pull request:
Canonicalizing OSSA provably minimizes the number of retains and
releases within the boundaries of that lifetime. This eliminates the
need for ad-hoc optimization of OSSA copies.

This initial implementation only canonicalizes owned values, but
canonicalizing guaranteed values is a simple extension.

This was originally part of the CopyPropagation prototype years
ago. Now OSSA is specified completely enough that it can be turned
into a simple utility instead.

CanonicalOSSALifetime uses PrunedLiveness to find the extended live
range and identify the consumes on the boundary. All other consumes
need their own copy. No other copies are needed.

By running this after other transformations that affect OSSA
lifetimes, we can avoid the need to run pattern-matching optimization
to SemanticARC to recover from suboptimal patterns, which is not
robust, maintainable, or efficient.

atrick committed:
Move SWIFT_ASSERT_ONLY to Compiler.h

atrick committed:
Add a CanonicalOSSALifetime utility.

Canonicalizing OSSA provably minimizes the number of retains and
releases within the boundaries of that lifetime. This eliminates the
need for ad-hoc optimization of OSSA copies.

This initial implementation only canonicalizes owned values, but
canonicalizing guaranteed values is a simple extension.

This was originally part of the CopyPropagation prototype years
ago. Now OSSA is specified completely enough that it can be turned
into a simple utility instead.

CanonicalOSSALifetime uses PrunedLiveness to find the extended live
range and identify the consumes on the boundary. All other consumes
need their own copy. No other copies are needed.

By running this after other transformations that affect OSSA
lifetimes, we can avoid the need to run pattern-matching optimization
to SemanticARC to recover from suboptimal patterns, which is not
robust, maintainable, or efficient.

gottesmm commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -216,14 +220,14 @@ OPERAND_OWNERSHIP(UnownedInstantaneousUse, UnmanagedAutoreleaseValue)
 OPERAND_OWNERSHIP(PointerEscape, ProjectBox) // The result is a T*.
 OPERAND_OWNERSHIP(PointerEscape, ProjectExistentialBox)
 OPERAND_OWNERSHIP(PointerEscape, UncheckedOwnershipConversion)
+OPERAND_OWNERSHIP(PointerEscape, ConvertEscapeToNoEscape)
```
The cvt escape to no escape change makes sense.

atrick replied:
Right, the general comment elsewhere that PointerEscape operands are basically bugs in the design that need to be fixed applies equally to convert_escape_to_no_escape. Obviously the original closure should have a well-defined scope.

gottesmm commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -261,6 +265,7 @@ OPERAND_OWNERSHIP(ForwardingBorrow, OpenExistentialValue)
 OPERAND_OWNERSHIP(ForwardingBorrow, OpenExistentialBoxValue)
 
 OPERAND_OWNERSHIP(EndBorrow, EndBorrow)
+OPERAND_OWNERSHIP(EndBorrow, EndApply)
```
I don't think this is correct. The end_apply just takes in the token from the begin_apply. What ownership semantics/type does a token have?

atrick replied:
The token only has ownership semantics. The whole reason it exists is to establish the scope of borrowed coroutine arguments. If it were considered a trivial use then it would be incorrectly ignored by any code that looks for uses relevant to the lifetime of borrowed arguments.

gottesmm replied:
Andy and I spoke about this offline. Can you add end_abort here as well and add what we talked about last night to this. I say something like this:

```
NOTE: Remembering that end_apply, abort_apply take a begin_apply token which has None ownership one may wonder why we would mark end_apply, abort_apply as an EndBorrow use. The reason why is that when one is walking def->use for ownership purposes, we will visit the end_apply, abort_apply use as a use of any values that we instantaneously borrow for the coroutine. In such a case, if we marked the use as trivial, we would lose semantic information.
```

I think that would clear up any questions this piece of code brings up.

atrick replied:
Explained in this commit
https://github.com/apple/swift/pull/35248/commits/07f5735e53bd340dc57ec16287ecc8d24f2b4ad3

gottesmm commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -399,20 +408,26 @@ OperandOwnershipClassifier::visitStoreBorrowInst(StoreBorrowInst *i) {
   // owned values may be passed to guaranteed arguments without an explicit
-  // borrow scope in the caller.
+  // borrow scope in the caller. A begin_apply, on the other hand, does have an
+  // explicit borrow scope in the caller.
```
Can you finish the sentence in this comment? My thought:

```
In contrast, a begin_apply /does/ have an explicit borrow scope in the caller so we must treat arguments passed to it as being borrowed for the entire borrow scope.
```

Also what about renaming  hasCallerScope to hasScopeInCaller? I think it may make it clear that you are not talking about something called a "caller scope" but more specifically something called a scope that is in the caller. But if you are against that I am ok with it.

gottesmm replied:
On another note, I was also thinking about this specific commit and I realized that someone reading it might be thinking that this might have been fixing a miscompile when I think that this is really just making the info more descriptive (the ownership verifier uses BorrowedOperand to discover the end of lifetime uses for borrowed operands, so we were always verifying this correctly). If you agree with my worry, can you add to this specific commit message that this is only enriching the info. Your thoughts?

atrick replied:
I applied your suggestions verbatim. These changes to operand ownership all fix bugs when we canonicalize OSSA. Previously we were only using it for verification though, so didn't need to be perfect for correctness.

atrick replied:
I'll fix the commit message.

gottesmm commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -435,7 +450,8 @@ OperandOwnershipClassifier::visitFullApply(FullApplySite apply) {
 
-  auto operandOwnership = getFunctionArgOwnership(argConv);
+  auto operandOwnership =
+    getFunctionArgOwnership(argConv, apply.beginsCoroutineEvaluation());
```
Can you describe the boolean arg with a /* like you did below?

gottesmm commented on include/swift/Basic/Compiler.h:
```diff
@@ -103,4 +103,12 @@
   "Please " SWIFT_BUG_REPORT_MESSAGE_BASE " and the crash backtrace."
 
+#ifdef NDEBUG
+#define SWIFT_ASSERT_ONLY_DECL(X)
```
Can you add a comment providing a description in words and an example of how this is supposed to be used?

atrick replied:
```
// Conditionally exclude declarations or statements that are only needed for
// assertions from release builds (NDEBUG) without cluttering the surrounding
// code by #ifdefs.
//
// struct DoThings  {
//   SWIFT_ASSERT_ONLY_DECL(unsigned verifyCount = 0);
//   DoThings() {
//     SWIFT_ASSERT_ONLY(verifyCount = getNumberOfThingsToDo());
//   }
//   void doThings() {
//     do {
//       // ... do each thing
//       SWIFT_ASSERT_ONLY(--verifyCount);
//     } while (!done());
//     assert(verifyCount == 0 && "did not do everything");
//   }
// };
```

gottesmm replied:
Thanks.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+/// the CanonicalOSSALifetime implementation.
+inline SILValue getCanonicalCopiedDef(SILValue v) {
+  while (true) {
+    if (auto *copy = dyn_cast<CopyValueInst>(v)) {
```
We already have a utility like this committed by Meghana. We shouldn't create these all over the place. Also I think you can do this cleaner with a while loop:

```
while (auto *copy = dyn_cast<CopyValueInst>(v))
  v = copy->getOperand();
return v;
```

atrick replied:
I won't use a separate utility here because, as the comment says, this implementation is coupled with the implementation inside `canonicalizeValueLifetime`. Other code that does not have such a dependence could use it though.

gottesmm replied:
Ok. I am fine with that.

gottesmm replied:
Can you hide it then and move it out of the global namespace?

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+///
+/// This result remains valid during copy rewriting. The only instructions
+/// referenced it contains are consumes that cannot be deleted.
+class CanonicalOSSAConsumeInfo {
```
Given that we may use this for borrowed values, can we excise the word "consume"?

atrick replied:
No, this is only "consume" information, regardless of whether the definition is owned or guaranteed.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+/// canonicalizeValueLifetime().
+struct CanonicalOSSALifetime {
+  // Current copied def for which this state describes the liveness.
+  SILValue currDef;
```
currDef -> currentDef.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+/// Allows the allocation of analysis state to be reused across calls to
+/// canonicalizeValueLifetime().
+struct CanonicalOSSALifetime {
+  // Current copied def for which this state describes the liveness.
```
doxygen 3 slashes

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+  // Current copied def for which this state describes the liveness.
+  SILValue currDef;
+
+  // Cumulatively, have any instructions been modified by canonicalization?
```
doxygen 3 slashes

atrick replied:
I don't understand why private members should have doc comments.

gottesmm replied:
They show up in doxygen.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+  // Pruned liveness for the extended live range including copies. For this
+  // purpose, only consuming instructions are considered "lifetime
+  // ending". end_borrows do not end a liverange that may include owned copies.
+  PrunedLiveness liveness;
```
doxygen 3 slashes

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+  PrunedLiveness liveness;
+
+  // Original points in the CFG where the current value's lifetime ends. This
+  // includes any point in which the value is consumed or destroyed. For
```
Same here.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+  //
+  // These blocks are not necessarily in the pruned live blocks since
+  // pruned liveness does not consider destroy_values.
+  llvm::SmallSetVector<SILBasicBlock *, 8> lifetimeEndBlocks;
```
You don't need llvm:: for this. We already import it into the swift namespace in Basic/LLVM.h.

gottesmm commented on lib/SILOptimizer/Transforms/CopyPropagation.cpp:
```diff
@@ -13,665 +13,28 @@
-/// Because this algorithm rewrites copies and destroys without attempting
-/// to balance the retain count, it is only sound for ownership-SSA.
+/// Because this algorithm rewrites copies and destroys without attempting to
+/// balance the retain count, it is only sound for ownership-SSA. The pass
```
`it is only sound for ownership-SSA` -> `it is only sound when SIL is in ownership-SSA form`.

gottesmm commented on lib/SILOptimizer/Transforms/CopyPropagation.cpp:
```diff
@@ -13,665 +13,28 @@
-/// TODO: Cleanup the resulting SIL. Delete instructions with no side effects
-/// that produce values which are immediately destroyed after copy propagation.
+/// TODO: Cleanup the resulting SIL by deleting instructions that produce dead
+/// values (after removing its copies).
```
Can you add another line here?

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,187 @@
+///
+/// Allows the allocation of analysis state to be reused across calls to
+/// canonicalizeValueLifetime().
+struct CanonicalOSSALifetime {
```
This name makes it sound like this is something related to the actual lifetime itself instead of the state used for Canonicalizing OSSA lifetimes. Why not something like: ``OSSALifetimeCanonicalizer``?

atrick replied:
This just represents the canonical OSSA lifetime before SIL transformation. That's all it is. It doesn't do anything.

gottesmm commented on lib/SILOptimizer/Transforms/CopyPropagation.cpp:
```diff
@@ -681,36 +44,33 @@ class CopyPropagation : public SILFunctionTransform {
   }
-
+  // Perform copy propgation for each copied value.
+  CanonicalOSSALifetime lifetime;
```
For instance (noting my comment about CanonicalOSSALifetime is confusing), I think this code would be clearer if it was this:

```
OSSALifetimeCanonicalizer canonicalizer;
```

or

```
OSSALifetimeCanonicalizationState canonicalizationState;
```

atrick replied:
I'll add State to the name to distinguish it from some hypothetical stand-alone lifetime representation

gottesmm commented on lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp:
```diff
@@ -0,0 +1,485 @@
+  SILInstruction *user = use->getUser();
+  SILBuilderWithScope B(user->getIterator());
+
+  auto loc = RegularLocation::getAutoGeneratedLocation(
```
@atrick is this same code as before just refactored out?

atrick replied:
At some point I probably changed this to use an auto-generated location. That and the tracing support are the only reasons it's factored out. Otherwise it's trivial.

atrick replied:
If you're asking what has changed about the algorithm in the past couple of years, or whenever, nothing except that
`computeCanonicalLiveness` now uses `OperandOwnership` which is now a fundamental verified part of OSSA.

gottesmm reviewed (commented):
Some questions/etc.

gottesmm commented on include/swift/SIL/SILValue.h:
```diff
@@ -626,9 +626,15 @@ llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    NonUse,
+
+    /// Uses that can only handle trivial values. The operand value must have
+    /// None ownership. These uses require liveness but are otherwise
```
This is a comment for the actual commit message. Please use a different term than undef for TrivialUse operand. It isn't undefined in a compiler sense of formatting your hard drive.

gottesmm replied:
Can you add a section to this commit that distinguishes in between trivial values and trivially typed things? I think that using the enum example we talked about with a caller would work well.

atrick replied:
I wrote a better commit message. Just to be clear, the ownership semantics of TrivialUse literally have undefined ownership semantics, which is the whole reason we want the compiler to enforce that they can't operate on values with ownership semantics (aka nontrivial values). Contrast this with things like InstantaneousUse or BitwiseEscape, which just don't care about ownership (i.e. they have no ownership semantics). We only get undefined behavior when the compiler does not enforce code that has undefined semantics.
https://github.com/apple/swift/pull/35248/commits/ce2a7cfe0edeb179589c99a2e2e9b0b90b06a819

atrick replied:
The enum cases shouldn't have much extra commentary so the enum itself is readable. But I clarified all the comments, and since "trivial value" isn't defined anywhere, I added an introduction in the OwnershipKind doc comment.
https://github.com/apple/swift/pull/35248/commits/ce2a7cfe0edeb179589c99a2e2e9b0b90b06a819

I find it grammatically impossible not to use the term "trivial values" when writing about ownership because the term "None values" doesn't parse. If I ever argued against using the word trivial, I deeply apologize for that.

gottesmm commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -776,14 +777,12 @@ OperandOwnership OperandOwnershipClassifier::visitBuiltinInst(BuiltinInst *bi) {
-  // anything and allows us to eliminate Optionals and thus confusion in between
-  // Optional::None and OwnershipKind::None.
+  // NOTE: NonUse distinguishes itself from InstantaneousUse because it does not
+  // require liveness. Discrimating such uses in the enum avoids the need to
```
This can be clearer. My thought:

```
NOTE: NonUse distinguishes itself from InstantaneousUse because it does not require liveness (e.x.: a use of a type dependent operand).
```

gottesmm commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -60,6 +60,11 @@ class OperandOwnershipClassifier
     return getOwnershipKind() == kind;
   }
 
+  OperandOwnership getBorrowOwnership();
```
Comments please.

gottesmm commented on lib/SIL/IR/OperandOwnership.cpp:
```diff
@@ -399,20 +408,27 @@ OperandOwnershipClassifier::visitStoreBorrowInst(StoreBorrowInst *i) {
-  // borrow scope in the caller.
+  // borrow scope in the caller. In contrast, a begin_apply /does/ have an
+  // explicit borrow scope in the caller so we must treat arguments passed to it
+  // as being borrowed for the entire borrow scope.
```
`for the entire borrow scope` -> `for the entire region of the coroutines execution that makes up that borrow scope`.

Or something like that.

gottesmm commented on include/swift/Basic/Compiler.h:
```diff
@@ -103,4 +103,29 @@
+//
+// struct DoThings  {
+//   SWIFT_ASSERT_ONLY_DECL(unsigned verifyCount = 0);
+//   DoThings() {
```
Nice comment!

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,188 @@
+///
+/// This top-level API rewrites the extended lifetime of a SILValue:
+///
+///     void canonicalizeValueLifetime(SILValue def, CanonicalOSSALifetime&)
```
`CanonicalOSSALifetime&` -> `CanonicalOSSALifetime &`

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalOSSALifetime.h:
```diff
@@ -0,0 +1,188 @@
+/// FIXME: Canonicalization currently bails out if any uses of the def has
+/// OperandOwnership::PointerEscape. Once project_box is protected by a borrow
+/// scope and mark_dependence is associated with an end_dependence,
+/// canonicalization will work everywhere as intended.
```
Can you add to this FIXME an explanation of the context why this was done? My thought add one more sentence to the end. `This was done so that we can forestall the need for more ad-hoc copy optimizations in the vast majority of cases where we do not have pointer escape uses.`

atrick committed:
Update and add unit tests for CopyPropagation coverage.

atrick committed:
Add a variation on CopyPropagation to handle debug_value correctly.

MandatoryCopyPropagation must be a separate pass in order to preserve
all debug_value instructions. CopyPropagation cannot preserve
debug_value because, as a rule, debug information cannot affect -O
behavior.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test linux platform

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test

atrick committed:
CanonicalOSSALifetime: canonicalize guaranteed values.

If a guaranteed value is not a recognized and handled borrow
introducer, then treat the copy as a separate owned live range.

atrick commented:
lldb has broken PR testing again
```
/home/buildnode/jenkins/workspace/swift-PR-Linux@2/branch-main/llvm-project/lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:2341:22: error: no member named 'make_scope_exit' in namespace 'llvm'
22:03:08   auto defer = llvm::make_scope_exit([&] {
22:03:08                ~~~~~~^
22:03:08 1 error generated.
```

atrick commented:
@swift-ci test
